nuagekubemon/client/nuageetcdclient.go

```diff
@@ -0,0 +1,624 @@
+			return nil
+		}
+		s := &etcdSubnetValue{ACTIVEIP: 0, VSDID: subnetInfo.ID, CIDR: subnetInfo.CIDR}
+		b, err = json.Marshal(s)
+		if err != nil {
+			glog.Errorf("json marshal struct(%v) failed", s, err)
+			return err
+		}
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Put(context.Background(), SUBNET_METADATA_TREE+ns+"/"+subnet, string(b))
+				if err != nil {
+					glog.Errorf("put on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		return nil
+	}, SUBNET_METADATA_TREE+"subnet-lock"); err != nil {
+		glog.Errorf("creating first subnet %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+		return err
+	}
+	return nil
+}
+
+//DeleteLastSubnet deletes last subnet in a namespace
+func (nuageetcd *NuageEtcdClient) DeleteLastSubnet(subnetInfo *api.EtcdSubnetMetadata) error {
+	var err error
+	var getResp *clientv3.GetResponse
+	ns := subnetInfo.Namespace
+	subnet := subnetInfo.Name
+	if err = nuageetcd.nuageMutex(func() error {
+		nuageEtcdRetry(
+			func() error {
+				getResp, err = nuageetcd.client.Get(context.Background(), SUBNET_METADATA_TREE+ns+"/"+subnet, clientv3.WithPrefix())
+				if err != nil {
+					glog.Errorf("get on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		if len(getResp.Kvs) == 0 {
+			return nil
+		}
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Delete(context.Background(), SUBNET_METADATA_TREE+ns+"/"+subnet, clientv3.WithPrefix())
+				if err != nil {
+					glog.Errorf("delete on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		return nil
+	}, SUBNET_METADATA_TREE+"subnet-lock"); err != nil {
+		glog.Errorf("deleting last subnet %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+		return err
+	}
+	return nil
+}
+
+//AllocateSubnet marks an entry if the subnet is not already used
+func (nuageetcd *NuageEtcdClient) AllocateSubnetCIDR(subnet *api.EtcdSubnetMetadata) (string, error) {
+	var err error
+	var getResp *clientv3.GetResponse
+	allocatedSubnet := ""
+	networkCIDR := strings.Replace(subnet.CIDR, "/", "-", -1)
+	err = nuageetcd.nuageMutex(func() error {
+		nuageEtcdRetry(
+			func() error {
+				getResp, err = nuageetcd.client.Get(context.Background(), POOL_CIDR_TREE, clientv3.WithPrefix())
+				if err != nil {
+					glog.Errorf("fetching all network cidrs from etcd failed: %v", err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		for _, kv := range getResp.Kvs {
+			if networkCIDR == path.Base(string(kv.Key)) {
+				allocatedSubnet = string(kv.Value)
+				return nil
+			}
+		}
+		nuageEtcdRetry(
+			func() error {
+				_, err = nuageetcd.client.Put(context.Background(), POOL_CIDR_TREE+networkCIDR, subnet.Name)
```

Comment:
What is the purpose of doing a etcd PUT call here after setting the allocatedSubnet above?